src/FakeItEasy/Creation/CastleDynamicProxy/CastleDynamicProxyGenerator.cs

```diff
@@ -122,17 +123,25 @@ private static ProxyGeneratorResult GetProxyResultForNoDefaultConstructor(Type t
         private static object DoGenerateProxy(
             Type typeOfProxy,
             ProxyGenerationOptions options,
-            IEnumerable<Type> additionalInterfacesToImplement,
+            ReadOnlyCollection<Type> additionalInterfacesToImplement,
             IEnumerable<object> argumentsForConstructor,
             IInterceptor interceptor)
         {
-            var allInterfacesToImplement = additionalInterfacesToImplement;
+            Type[] allInterfacesToImplement;
 
             if (typeOfProxy.GetTypeInfo().IsInterface)
             {
-                allInterfacesToImplement = new[] { typeOfProxy }.Concat(allInterfacesToImplement);
+                allInterfacesToImplement = new Type[1 + additionalInterfacesToImplement.Count];
+                additionalInterfacesToImplement.CopyTo(allInterfacesToImplement, 1);
+
+                allInterfacesToImplement[0] = typeOfProxy;
                 typeOfProxy = typeof(object);
             }
+            else
+            {
+                allInterfacesToImplement = new Type[additionalInterfacesToImplement.Count];
+                additionalInterfacesToImplement.CopyTo(allInterfacesToImplement, 0);

```

Comment:
Could you perhaps pass `additionalInterfacesToImplement` as an array rather than `ReadOnlyCollection<Type>`? It would avoid this copy.

Reply:
Happy to. I used a `ReadOnlyCollection` so the values couldn't be mutated, but I am more than happy to trust our internal code not to do so.  I will make the change.

Reply:
Oh, I wasn't thinking earlier this morning. The interfaces are added one by one to `ProxyOptions`, so we're accumulating in a flexible list. If I pass into the proxy generator as an array, I'll have to copy to an array earlier anyhow, so it's the same amount of work. If you think it would look better here, I can still do so, but I'd be happier to leave as is, since we're not increasing the amount of work and we retain readonlyness throughout.

Reply:
Ah, I see. Fine by me, then.